core/src/main/java/org/elasticsearch/common/util/IndexFolderUpgrader.java

```diff
@@ -0,0 +1,155 @@
+            }
+        }
+    }
+
+    /**
+     * Renames <code>indexFolderName</code> index folders found in node paths and custom path
+     * iff {@link #needsUpgrade(Index, String)} is true.
+     * Index folder in custom paths are renamed first followed by index folders in each node path.
+     */
+    void upgrade(final String indexFolderName) throws IOException {
+        for (NodeEnvironment.NodePath nodePath : nodeEnv.nodePaths()) {
+            final Path indexFolderPath = nodePath.indicesPath.resolve(indexFolderName);
+            if (Files.exists(indexFolderPath) == false) {
+                continue;
+            }
+            final IndexMetaData indexMetaData = indexStateFormat.loadLatestState(logger, indexFolderPath);
+            if (indexMetaData != null) {
+                final Index index = indexMetaData.getIndex();
+                if (needsUpgrade(index, indexFolderName)) {
+                    logger.info("{} upgrading [{}] to new naming convention", index, indexFolderPath);
+                    final IndexSettings indexSettings = new IndexSettings(indexMetaData, settings);
+                    if (indexSettings.hasCustomDataPath()) {
+                        // we rename index folder in custom path before renaming them in any node path
+                        // to have the index state under a not-yet-upgraded index folder, which we use to
+                        // continue renaming after a incomplete upgrade.
+                        // NOTE: we rename the folder in custom path only the first time we encounter a state
+                        // in a node path, which needs upgrading, it is a no-op for subsequent node paths
+                        upgrade(index, nodeEnv.resolveCustomLocation(indexSettings).resolve(indexFolderName));
+                    }
+                    upgrade(index, indexFolderPath);
+                } else {
+                    logger.info("[{}] no upgrade needed - already upgraded", indexFolderPath);
+                }
+            } else {
+                logger.info("[{}] no index state found - ignoring", indexFolderPath);
+            }
+        }
+    }
+
+    /**
+     * Upgrades all indices found under <code>nodeEnv</code>. Already upgraded indices are ignored.
+     */
+    public static void upgradeIndicesIfNeeded(final Settings settings, final NodeEnvironment nodeEnv) throws IOException {
+        final IndexFolderUpgrader upgrader = new IndexFolderUpgrader(settings, nodeEnv);
+        for (String indexFolderName : nodeEnv.availableIndexFolders()) {
+            upgrader.upgrade(indexFolderName);
+        }
+    }
+
+    static boolean needsUpgrade(Index index, String indexFolderName) {
+        return indexFolderName.equals(index.getUUID()) == false;
+    }
+
+    static MetaDataStateFormat<IndexMetaData> metaDataMetaDataStateFormat() {
+        return new MetaDataStateFormat<IndexMetaData>(XContentType.SMILE, MetaStateService.INDEX_STATE_FILE_PREFIX) {
```

Comment:
can we add a comment that this XContentType is not used? (and we should really remove the config option to change it :) - different change)